plat/allwinner/common/sunxi_common.c

```diff
@@ -156,3 +158,52 @@ int platform_init_r_twi(uint16_t socid, bool use_i2c)
 
 	return 0;
 }
+
+/* This lock synchronises access to the arisc management processor. */
+DEFINE_BAKERY_LOCK(arisc_lock);
+
+/*
+ * Tell the "arisc" SCP core (an OpenRISC core) to execute some code.
+ * We don't have any service running there, so we place some OpenRISC code
+ * in SRAM, put the address of that into the reset vector and release the
+ * arisc reset line. The SCP will execute that code and pull the line up again.
+ */
+void sunxi_execute_arisc_code(uint32_t *code, size_t size,
+			      int patch_offset, uint16_t param)
+{
+	uintptr_t arisc_reset_vec = SUNXI_SRAM_A2_BASE - 0x4000 + 0x100;
+
+	do {
+		bakery_lock_get(&arisc_lock);
+		/* Wait until the arisc is in reset state. */
+		if (!(mmio_read_32(SUNXI_R_CPUCFG_BASE) & BIT(0)))
+			break;
+
+		bakery_lock_release(&arisc_lock);
+	} while (1);
+
+	/* Patch up the code to feed in an input parameter. */
+	if (patch_offset >= 0 && patch_offset <= (size - 4))
+		code[patch_offset] = (code[patch_offset] & ~0xffff) | param;
+	flush_dcache_range((uintptr_t)code, size);
```

Comment:
Strictly speaking you don't need to flush, clean is enough. Same below.

Reply:
True, good catch. Which makes me think whether we need to do something in ATF about Cortex-A53 erratum 855873 (promoting cache cleans to invalidate in code for cores before r0p3).

Reply:
It looks like we are safe as part of the workaround for 82631. However, we should make it clearer, or make it explicit in the workaround for 855873.